cub/cub/agent/agent_topk.cuh

```diff
@@ -0,0 +1,835 @@
+
+  //---------------------------------------------------------------------
+  // Constructor
+  //---------------------------------------------------------------------
+  /**
+   * @param temp_storage
+   *   Reference to temp_storage
+   *
+   * @param d_keys_in
+   *   Input data, keys
+   *
+   * @param d_keys_out
+   *   Output data, keys
+   *
+   * @param d_values_in
+   *   Input data, values
+   *
+   * @param d_values_out
+   *   Output data, values
+   *
+   * @param num_items
+   *   Total number of input items
+   *
+   * @param k
+   *   The K value. Will find K elements from num_items elements
+   *
+   * @param extract_bin_op
+   *   Extract bin operator
+   *
+   * @param identify_candidates_op
+   *   Filter operator
+   *
+   */
+  _CCCL_DEVICE _CCCL_FORCEINLINE AgentTopK(
+    TempStorage& temp_storage,
+    const KeyInputIteratorT d_keys_in,
+    KeyOutputIteratorT d_keys_out,
+    const ValueInputIteratorT d_values_in,
+    ValueOutputIteratorT d_values_out,
+    OffsetT num_items,
+    OutOffsetT k,
+    ExtractBinOpT extract_bin_op,
+    IdentifyCandidatesOpT identify_candidates_op)
+      : temp_storage(temp_storage.Alias())
+      , d_keys_in(d_keys_in)
+      , d_keys_out(d_keys_out)
+      , d_values_in(d_values_in)
+      , d_values_out(d_values_out)
+      , num_items(num_items)
+      , k(k)
+      , extract_bin_op(extract_bin_op)
+      , identify_candidates_op(identify_candidates_op)
+  {}
+
+  //---------------------------------------------------------------------
+  // Utility methods for device topK
+  //---------------------------------------------------------------------
+  /**
+   * Try to perform vectorized data loading.
+   */
+  template <typename Func, typename T>
+  _CCCL_DEVICE _CCCL_FORCEINLINE void ConsumeRange(T in, const OffsetT num_items, Func f)
+  {
+    key_in_t thread_data[ITEMS_PER_THREAD];
+
+    const OffsetT ITEMS_PER_PASS   = TILE_ITEMS * gridDim.x;
+    const OffsetT total_num_blocks = (num_items - 1) / TILE_ITEMS + 1;
+
+    const OffsetT num_remaining_elements = num_items % TILE_ITEMS;
+    const OffsetT last_block_id          = (total_num_blocks - 1) % gridDim.x;
+
+    OffsetT tile_base = blockIdx.x * TILE_ITEMS;
+    OffsetT offset    = threadIdx.x * ITEMS_PER_THREAD + tile_base;
+
+    for (int i_block = blockIdx.x; i_block < total_num_blocks - 1; i_block += gridDim.x)
+    {
+      block_load_input_t(temp_storage.load_input).Load(in + tile_base, thread_data);
+      for (int j = 0; j < ITEMS_PER_THREAD; ++j)
+      {
+        f(thread_data[j], offset + j);
+      }
+      tile_base += ITEMS_PER_PASS;
+      offset += ITEMS_PER_PASS;
+    }
+
+    if (blockIdx.x == last_block_id)
+    {
+      if (num_remaining_elements == 0)
+      {
+        block_load_input_t(temp_storage.load_input).Load(in + tile_base, thread_data);
+      }
+      else
+      {
+        block_load_input_t(temp_storage.load_input).Load(in + tile_base, thread_data, num_remaining_elements, 0);
```

Comment:
important: Assigning `0` to the out-of-bounds items will prevent using key types that are not `0`-assignable. Do we rely on those items being `0`? Otherwise, we can just assigning a default value. 
```suggestion
        block_load_input_t(temp_storage.load_input).Load(in + tile_base, thread_data, num_remaining_elements);
```